### PR TITLE
feat: implement test sharding

### PR DESCRIPTION
**Pull request by thesayyn**

### CODE REVIEW

File: jasmine/defs.bzl [hunk 1]
```diff
@@ -23,5 +23,6 @@ def jasmine_test(jasmine_repository = "jasmine", **kwargs):
             "@{}//:node_modules/jasmine".format(jasmine_repository),
             "@{}//:node_modules/jasmine-core".format(jasmine_repository),
         ],
+        chdir = kwargs.pop("chdir", native.package_name()),
```
Comment (thesayyn) [inline jasmine/defs.bzl:26]: This slipped away while handling merge conflicts earlier PRs 💆

File: jasmine/private/runner.js [hunk 2]
```diff
@@ -1,4 +1,65 @@
...
+            constructor(options) {
+                super(options);
+                getTotalSpecsDefined = () => this.totalSpecsDefined;
+            }
+        };
+    }
+
+    const Order = jasmine.Order;
+    jasmine.Order = ($j) => {
+        const RandomOrder = Order($j);
```
Comment (jbedard) [inline jasmine/private/runner.js:33]: Why "RandomOrder" instead of "DefaultOrder"? Do we know it's random?

File: jasmine/private/runner.js [hunk 3]
```diff
@@ -1,4 +1,65 @@
...
+            constructor(options) {
+                this.random = new RandomOrder(options);
+            }
+            sort(items) {
+                const totalSpecsDefined = getTotalSpecsDefined();
+                const minIndex = (totalSpecsDefined * shardIndex) / shardCount;
+                const maxIndex = (totalSpecsDefined * (shardIndex + 1)) / shardCount;
+
+                const filtered = [];
+                for (const item of items) {
```
Comment (jbedard) [inline jasmine/private/runner.js:44]: Any nicer doing `items.filter(...)`?

**Comment by jbedard** [comment]
This seems quite different then how rules_nodejs did it. Was there a reason for doing it differently?

File: jasmine/tests/failing_shard.test.js [hunk 4]
```diff
@@ -0,0 +1,8 @@
+describe('test failures with sharding', () => {
```
Comment (jbedard) [inline jasmine/tests/failing_shard.test.js:1]: Should only 1/2 of the shards fail?

File: jasmine/private/runner.js (hunk 2 above)
Comment (thesayyn) [reply jasmine/private/runner.js:33]: Good point. DefaultOrder is reasonable

File: jasmine/tests/shard.test.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
+describe('test sharding with explicit isolation failures', () => {
```
Comment (jbedard) [inline jasmine/tests/shard.test.js:1]: which "failures"? 🤔

**Comment by thesayyn** [comment]
> This seems quite different then how rules_nodejs did it. Was there a reason for doing it differently?

rules_nodejs does not use `jasmine` package at all. it only depends on `jasmine-core`

**Comment by jbedard** [comment]
> rules_nodejs does not use jasmine package at all. it only depends on jasmine-core

What are the advantages to how you implemented workers here vs rules_nodejs?

**Comment by jbedard** [comment]
Do we need to handle the different exit codes like rules_nodejs does? You've used `TESTS_FAILED`, what about `NO_TESTS_FOUND`. rules_nodejs also uses `INTERRUPTED` although I'm not sure when that would happen 🤔 

https://github.com/bazelbuild/bazel/blob/507d85fd6dcf276add3b46158eb5bf36a4179db5/src/main/java/com/google/devtools/build/lib/util/ExitCode.java#L49

File: jasmine/tests/failing_shard.test.js (hunk 4 above)
Comment (thesayyn) [reply jasmine/tests/failing_shard.test.js:1]: nope. they both fail.

Comment (jbedard) [reply jasmine/tests/failing_shard.test.js:1]: I mean making only 1 shard fail and ensuring bazel still considers that a failure, not only when every shard fails...

File: jasmine/tests/shard.test.js (hunk 5 above)
Comment (jbedard) [reply jasmine/tests/shard.test.js:1]: That's how I understood it when I read the code, but the description here mentions "failures" and I'm not sure what those failures are..?

**Comment by thesayyn** [comment]
> > rules_nodejs does not use jasmine package at all. it only depends on jasmine-core
> 
> What are the advantages to how you implemented workers here vs rules_nodejs?

1- less code to maintain. less API surface that we are depending on
2- it is the official jasmine node runner
3- it handles premature exists https://github.com/jasmine/jasmine-npm/blob/2884a0c986ae05647c15f4e4db40905b09a52692/lib/jasmine.js#L470

**Comment by thesayyn** [comment]
> Do we need to handle the different exit codes like rules_nodejs does? You've used `TESTS_FAILED`, what about `NO_TESTS_FOUND`. rules_nodejs also uses `INTERRUPTED` although I'm not sure when that would happen 🤔
> 
> https://github.com/bazelbuild/bazel/blob/507d85fd6dcf276add3b46158eb5bf36a4179db5/src/main/java/com/google/devtools/build/lib/util/ExitCode.java#L49

I am not sure if it is worth maintain bunch of launcher code just for this. https://github.com/jasmine/jasmine-npm/blob/2884a0c986ae05647c15f4e4db40905b09a52692/lib/jasmine.js#L541

File: jasmine/private/runner.js [hunk 6]
```diff
@@ -1,4 +1,65 @@
 const path = require("path");
-const jasmine_path = require.resolve("jasmine");
 
-require(path.join(jasmine_path, "..", "..", "bin", "jasmine.js"));
+const shardCount = Number(process.env.TEST_TOTAL_SHARDS);
+const shardIndex = Number(process.env.TEST_SHARD_INDEX);
+const shardStatusFile = process.env.TEST_SHARD_STATUS_FILE;
```
Comment (jbedard) [inline jasmine/private/runner.js:5]: Interesting... 🤔 

I've been working with sharding via rules_nodejs `nodejs_test` for a while now and never seen this. Do you have more info other then the https://bazel.build/reference/test-encyclopedia#test-sharding ?

File: jasmine/tests/failing_shard.test.js (hunk 4 above)
Comment (thesayyn) [reply jasmine/tests/failing_shard.test.js:1]: Ah. i just took this test from rules_nodejs.  We could add that case later.

File: jasmine/private/runner.js (hunk 3 above)
Comment (thesayyn) [reply jasmine/private/runner.js:44]: I find it uglier 🥲

File: jasmine/private/runner.js (hunk 6 above)
Comment (thesayyn) [reply jasmine/private/runner.js:5]: This goes by round-robin partitioning in SQL world but a slightly different version of it. I found this https://www.researchgate.net/figure/The-four-partitioning-strategies-a-round-robin-b-range-c-list-and-d-hash_fig1_276075813

Comment (thesayyn) [reply jasmine/private/runner.js:5]: Though it took the code from rules_jest.

Comment (jbedard) [reply jasmine/private/runner.js:5]: Sorry I specially mean the `TEST_SHARD_STATUS_FILE` status file...

Comment (alexeagle) [reply jasmine/private/runner.js:5]: If you don't touch the file, Bazel will think the test runner doesn't support sharping

Comment (jbedard) [reply jasmine/private/runner.js:5]: Yeah that's what the docs say (the link above), but that's all it says.

Doing `nodejs_test(shard_count=N)` seems to work fine without touching this file. Does `nodejs_test` do something I'm not seeing?

With `nodejs_test` I'm consuming the 2 sharding variables and do the sharding very similar to this PR and it seems to be working fine. It always launches with N processes, the processes divide the specs manually etc. I guess I should try this and see if it changes anything...? 🤔

File: jasmine/tests/shard.test.js (hunk 5 above)
Comment (thesayyn) [reply jasmine/tests/shard.test.js:1]: yeah, i am not sure either. open to suggestions.

Comment (thesayyn) [reply jasmine/tests/shard.test.js:1]: changed it to "assert sharding runs specs(it blocks) in the expected order and partitions equally over the available shards"

**Review by jbedard — APPROVED** [review]

**Pull request merged**